firedrake/mg/interface.py

```diff
@@ -1,106 +1,132 @@
+
+    c2f_map = utils.coarse_to_fine_node_map(cfs, ffs)
+    weights = utils.get_restriction_weights(cfs, ffs)
 
     args = [coarse.dat(op2.INC, coarse.cell_node_map()[op2.i[0]]),
-            fine.dat(op2.READ, hierarchy.cell_node_map(lvl))]
+            fine.dat(op2.READ, c2f_map)]
 
-    if not hierarchy._discontinuous:
-        weight = weights[lvl+1]
-        args.append(weight.dat(op2.READ, hierarchy.cell_node_map(lvl)))
+    if weights is not None:
+        args.append(weights.dat(op2.READ, c2f_map))
     coarse.dat.zero()
-    op2.par_loop(hierarchy._restrict_kernel, hierarchy._cell_sets[lvl],
-                 *args)
+    op2.par_loop(kernel, c2f_map.iterset, *args)
 
 
 @firedrake.utils.known_pyop2_safe
 def inject(fine, coarse):
     cfs = coarse.function_space()
-    hierarchy, lvl = utils.get_level(cfs)
+    ffs = fine.function_space()
+    hierarchy, lvl = utils.get_level(cfs.mesh())
     if hierarchy is None:
         raise RuntimeError("Coarse function not from hierarchy")
-    fhierarchy, flvl = utils.get_level(fine.function_space())
+    fhierarchy, flvl = utils.get_level(ffs.mesh())
     if flvl != lvl + 1:
         raise ValueError("Can only inject from level %d to level %d, not %d" %
                          (flvl, flvl - 1, lvl))
     if hierarchy is not fhierarchy:
         raise ValueError("Can't prolong between functions from different hierarchies")
-    if isinstance(hierarchy, firedrake.MixedFunctionSpaceHierarchy):
+    if len(cfs) > 1:
+        assert len(ffs) == len(cfs)
         for f, c in zip(fine.split(), coarse.split()):
             inject(f, c)
         return
-    op2.par_loop(hierarchy._inject_kernel, hierarchy._cell_sets[lvl],
+    kernel = utils.get_transfer_kernel(cfs, ffs, typ="inject")
+    c2f_map = utils.coarse_to_fine_node_map(cfs, ffs)
+    op2.par_loop(kernel, c2f_map.iterset,
                  coarse.dat(op2.WRITE, coarse.cell_node_map()[op2.i[0]]),
-                 fine.dat(op2.READ, hierarchy.cell_node_map(lvl)))
+                 fine.dat(op2.READ, c2f_map))
+
+
+def FunctionHierarchy(fs_hierarchy, functions=None):
+    """ outdated and returns warning & list of functions corresponding to each level
+    of a functionspace hierarchy
+
+        :arg fs_hierarchy: the :class:`~.FunctionSpaceHierarchy` to build on.
+        :arg functions: optional :class:`~.Function` for each level.
+
+    """
+    from firedrake.logging import warning, RED
+    warning(RED % "FunctionHierarchy is obsolete. Falls back by returning list of functions")
+
+    if functions is not None:
+        assert len(functions) == len(fs_hierarchy)
+        for f, V in zip(functions, fs_hierarchy):
+            assert f.function_space() == V
+        return tuple(functions)
+    else:
+        return tuple([firedrake.Function(f) for f in fs_hierarchy])
+
+
+def FunctionSpaceHierarchy(mesh_hierarchy, *args, **kwargs):
+    return tuple(firedrake.FunctionSpace(mesh, *args, **kwargs) for mesh in mesh_hierarchy)
```

Comment:
Could we have some warnings in here to remind user of the depreciation, at least until I've had a chance to remove uses of FSH in some of my firedrake-mlmc stuff?